system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the Ignore Case Backend of the new file system.
+// It is a virtual backend that normalizes all file paths to lower case.
+
+#include "backend.h"
+#include "file.h"
+#include "memory_backend.h"
+#include "wasmfs.h"
+
+namespace {
+
+std::string normalize(const std::string& name) {
+  std::string result = name;
+  for (auto& ch : result) {
+    ch = std::tolower(ch);
+  }
+  return result;
+}
+} // namespace
+
+namespace wasmfs {
+
+// Problem: Child entries are stored both in IgnoreCaseDirectory and in baseDirectory.
+//
+// Original name case preservation could be possible if MemoryDirectory::entries
+// were accesible. Then, we would store the original case there but search
+// ignoring case.
```

Comment:
Let's just go ahead and make everything in `MemoryDirectory` public. If we're going to use it as a general utility, then we should lean into that.

Reply:
I'd suggest to work on case preservation separately. This PR is pretty big and takes too long already.